File: solution/무인열차/무인열차_solution.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
	int N; // 세로크기
	int M; // 가로크기
	int map[][] = new int [50 + 10][50 + 10]; // 사진 데이터
	int count;
	class DATA{
        int r, c, t;
        DATA(int r, int c){
        	this.r=r; this.c=c;
        }
    }
	DATA dat[][] = new DATA[2][50*50];

	void InputData(){
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		M = sc.nextInt();
		
		for (int i = 1; i<=N; i++){
			for (int j = 1; j<=M; j++){
				map[i][j]= sc.nextInt();
			}
		}
		sc.close();
	}
	void Bfs(int r, int c, int p){
		int dr[] = {0,0,1,-1};
		int dc[] = {1,-1,0,0};
		Queue <DATA> que = new LinkedList <DATA>();
		que.offer(new DATA(r,c));
		dat[p][count] = new DATA(r,c);
		
		while(!que.isEmpty()) {
			DATA d = que.poll();
			for(int i=0; i<4; i++){
				int nr = d.r + dr[i];
				int nc = d.c + dc[i];
				if (map[nr][nc] == 0) continue;
				map[nr][nc] = 0;
				count++;
				que.offer(new DATA(nr,nc));
				dat[p][count] = new DATA(nr,nc);
			}
		}
	}
	int Solve(){
		int ans = 0x7FFFFFFF;
		int point = 0;
	    int cnt[]=new int[2];

		for (int i = 1; i <= N; i++){
	        for (int j = 1; j <= M; j++){
	            if (map[i][j] == 1){
	                count = 1;
	                Bfs(i, j, point);
	                cnt[point] = count;
	                point++;
	            }
	        }
	    }
	    for (int i = 1; i <= cnt[0]; i++){
	        for (int j = 1; j <= cnt[1]; j++){
	            int temp = Math.abs(dat[0][i].r - dat[1][j].r) + Math.abs(dat[0][i].c - dat[1][j].c);
	            if (temp < ans) ans = temp;
	        }
	    }
	    return ans - 1;
	}
	public static void main(String[] args){
		Main m = new Main();

		m.InputData();	 //	입력 함수

		//	코드를 작성하세요
		int ans = m.Solve();
		System.out.println(ans);
	}
}

File: solution/무인열차/무인열차_solution.c
#include <stdio.h>

int N; // 세로크기
int M; // 가로크기
int map[50 + 10][50 + 10]; // 사진 데이터

struct st
{
    int r, c;
}dat[2][50*50];
struct st queue[50*50];
int rp, wp;
int count;
int dr[] = {0,0,1,-1};
int dc[] = {1,-1,0,0};

#define ABS(a) (((a)<0) ? -(a) : (a))

void Bfs(int r, int c, int p){
	int i, nr, nc;
	struct st d;
	wp = rp = 0;

	dat[p][count].r = queue[wp].r = r;
	dat[p][count].c = queue[wp].c = c;
	wp++;
	while(wp>rp) {
		d = queue[rp++];
		for(i=0; i<4; i++){
			nr = d.r + dr[i];
			nc = d.c + dc[i];
			if (map[nr][nc] == 0) continue;
			map[nr][nc] = 0;
			count++;
			dat[p][count].r = queue[wp].r = nr;
			dat[p][count].c = queue[wp].c = nc;
			wp++;
		}
	}
}
int Solve(void){
	int i, j, temp, ans = 0x7FFFFFFF;
	int point = 0;
    int cnt[2];

	for (i = 1; i <= N; i++){
        for (j = 1; j <= M; j++){
            if (map[i][j] == 1){
                count = 1;
                Bfs(i, j, point);
                cnt[point] = count;
                point++;
            }
        }
    }
    for (i = 1; i <= cnt[0]; i++){
        for (j = 1; j <= cnt[1]; j++){
            temp = ABS(dat[0][i].r - dat[1][j].r) + ABS(dat[0][i].c - dat[1][j].c);
            if (temp < ans) ans = temp;
        }
    }
    return ans - 1;
}

void InputData(void){
	int i, j;
	scanf("%d %d", &N, &M);
	for (i = 1; i <= N; i++){
		for (j = 1; j <= M; j++){
			scanf("%d", &map[i][j]);
		}
	}
}

int main(void){
	int ans;
	
	InputData(); // 입력함수
	
	// 코드를 작성 하세요
	ans = Solve();
    printf("%d", ans);
    return 0;
}